src/common/io/http/response.c

```diff
@@ -0,0 +1,412 @@
+    uint64_t contentSize;                                           // Content size (ignored for chunked)
+    uint64_t contentRemaining;                                      // Content remaining (per chunk if chunked)
+    bool closeOnContentEof;                                         // Will server close after content is sent?
+    bool contentExists;                                             // Does content exist?
+    bool contentEof;                                                // Has all content been read?
+    Buffer *content;                                                // Caches content once requested
+};
+
+OBJECT_DEFINE_MOVE(HTTP_RESPONSE);
+OBJECT_DEFINE_FREE(HTTP_RESPONSE);
+
+OBJECT_DEFINE_GET(IoRead, , HTTP_RESPONSE, IoRead *, contentRead);
+OBJECT_DEFINE_GET(Code, const, HTTP_RESPONSE, unsigned int, code);
+OBJECT_DEFINE_GET(Header, const, HTTP_RESPONSE, const HttpHeader *, header);
+OBJECT_DEFINE_GET(Reason, const, HTTP_RESPONSE, const String *, reason);
+
+/***********************************************************************************************************************************
+When response is done close/reuse the connection
+***********************************************************************************************************************************/
+static void
+httpResponseDone(HttpResponse *this)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(HTTP_RESPONSE, this);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(this->session != NULL);
+
+    // If close was requested by the server then free the session
+    if (this->closeOnContentEof)
+    {
+        httpSessionFree(this->session);
+
+        // Only update the close stats after a successful response so it is not counted if there was an error/retry
+        httpClientStat.close++;
+    }
+    // Else return it to the client do it can be reused
+    else
+        httpSessionDone(this->session);
+
+    this->session = NULL;
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/***********************************************************************************************************************************
+Read content
+***********************************************************************************************************************************/
+static size_t
+httpResponseRead(THIS_VOID, Buffer *buffer, bool block)
+{
+    THIS(HttpResponse);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(HTTP_RESPONSE, this);
+        FUNCTION_LOG_PARAM(BUFFER, buffer);
+        FUNCTION_LOG_PARAM(BOOL, block);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(buffer != NULL);
+    ASSERT(!bufFull(buffer));
+    ASSERT(this->contentEof || this->session != NULL);
+
+    // Read if EOF has not been reached
+    size_t actualBytes = 0;
+
+    if (!this->contentEof)
+    {
+        IoRead *rawRead = httpSessionIoRead(this->session);
```

Comment:
So I guess everything created in this function is part of the memContext of the calling function (except line 135)?

Reply:
That's the only place that memory is allocated in this function, so yes. Of course, that could change.

When this code was originally written (it was moved from httpClient in this commit) mem context blocks were much more expensive because they contained a TRY...FINALLY. Now that they are cheaper there seems to be little harm in wrapping the entire block in a temp mem context. So done.